src/dhcp/clientv4.rs

```diff
@@ -0,0 +1,425 @@
+use {Result, Error};
+use wire::{IpVersion, IpProtocol, IpCidr, IpEndpoint, IpAddress,
+           Ipv4Cidr, Ipv4Address, Ipv4Packet, Ipv4Repr,
+           UdpPacket, UdpRepr,
+           DhcpPacket, DhcpRepr, DhcpMessageType};
+use wire::dhcpv4::field as dhcpv4_field;
+use socket::{SocketSet, SocketHandle, RawSocket, RawSocketBuffer};
+use phy::{Device, ChecksumCapabilities};
+use iface::EthernetInterface as Interface;
+use time::{Instant, Duration};
+use super::{UDP_SERVER_PORT, UDP_CLIENT_PORT};
+
+const DISCOVER_TIMEOUT: u64 = 10;
+const REQUEST_TIMEOUT: u64 = 1;
+const REQUEST_RETRIES: u16 = 15;
+const RENEW_INTERVAL: u64 = 60;
+const RENEW_RETRIES: u16 = 3;
+const PARAMETER_REQUEST_LIST: &[u8] = &[
+    dhcpv4_field::OPT_SUBNET_MASK,
+    dhcpv4_field::OPT_ROUTER,
+    dhcpv4_field::OPT_DOMAIN_NAME_SERVER,
+];
+
+#[derive(Debug)]
+struct RequestState {
+    retry: u16,
+    server: Ipv4Address,
+}
+
+#[derive(Debug)]
+struct RenewState {
+    retry: u16,
+    server: Ipv4Address,
+}
+
+#[derive(Debug)]
+enum ClientState {
+    /// Discovering the DHCP server
+    Discovering,
+    /// Requesting an address
+    Requesting(RequestState),
+    /// Having an address, refresh it periodically
+    Renew(RenewState),
+}
+
+pub struct Client {
+    state: ClientState,
+    raw_handle: SocketHandle,
+    /// When to send next request
+    next_egress: Instant,
+    transaction_id: u32,
+    dns_servers: [Option<Ipv4Address>; 3]
+}
+
+/// DHCP client with a RawSocket.
+///
+/// To provide memory for the dynamic IP address, configure your
+/// `Interface` with one of `ip_addrs` and the `ipv4_gateway` being
+/// `Ipv4Address::UNSPECIFIED`.
+///
+/// You must call `poll()` after `iface.poll()` to send and receive
+/// DHCP packets.
+impl Client {
+    /// TODO
+    pub fn new<'a, 'b, 'c>(sockets: &mut SocketSet<'a, 'b, 'c>, rx_buffer: RawSocketBuffer<'b, 'c>, tx_buffer: RawSocketBuffer<'b, 'c>, now: Instant) -> Self
+    where 'b: 'c,
+    {
+        let raw_socket = RawSocket::new(IpVersion::Ipv4, IpProtocol::Udp, rx_buffer, tx_buffer);
+        let raw_handle = sockets.add(raw_socket);
+
+        Client {
+            state: ClientState::Discovering,
+            raw_handle,
+            next_egress: now,
+            transaction_id: 1,
+            dns_servers: [None; 3],
+        }
+    }
+
+    /// When to send next packet
+    ///
+    /// Useful for suspending execution after polling.
+    pub fn next_poll(&self, now: Instant) -> Duration {
+        self.next_egress - now
+    }
+
+    /// Process incoming packets on the contained RawSocket, and send
+    /// DHCP requests when timeouts are ready.
+    pub fn poll<DeviceT: for<'d> Device<'d>>(&mut self, iface: &mut Interface<DeviceT>, sockets: &mut SocketSet, now: Instant) -> Result<()> {
```

Comment:
Not returning anything here means that this code can't be used for other types of host configuration (notably: [PXE](http://www.pix.net/software/pxeboot/archive/pxespec.pdf), for network booting).  DHCP is designed to do more than just networking parameters.  Is there some reason not to return some representation of the response packet?

Reply:
I have given it a try to return the `DhcpRepr<'a>` but that is really cumbersome as it is bound to the lifetime of the packet received by `raw_socket.recv()` in this function. Check my branch `dhcp-return` if you would like to work on that.

Proposals are welcome. This is just a as-simple-as-possible implementation.

Reply:
Step one I'll get a proof of concept working locally (otherwise my proposal is likely to miss some important details).   I'll be back.

Reply:
any news?

Reply:
No, I got put onto a different thing... :-/

Reply:
Did you say here that there is genuine commercial interest in smoltcp and embedded Rust?

Reply:
Why wouldn't there be?